### DISCUSSION TITLE
Previous savedata can be converted to a .dat.backup while the new savedata is saving.

### DISCUSSION THREAD

**Discussion by Jdance-Media** [SmartlyDressedGames/Unturned-3.x-Community#4636] [Ideas]
After the whole incident on the preview branch, I, among other plugin developers, came to the realization that in the case that writing to the file is somehow stopped midway through, the previous savedata file should be renamed to a .dat.backup prior to saving so that it can be used rather than losing the data.

**Comment by SDGNelson** [comment]
This might actually be relatively straightforward to implement by adding the file move in `ServerSavedata` write methods? Would it be useful to have the previous version of the file even if the game doesn't have a way to revert to it?

**Reply by Jdance-Media** [reply]
I personally feel as if it will be helpful to have at least some working copy of the file, even if using it requires a bit more work than the game just defaulting to it. It will be easy enough instruct users on my end to just rename the file if it needs to be used.

In reality it would be best if errors were caught when loading the file and it defaulted to the backup if available, but that isn't something that can be added as easily.

**Reply by SDGNelson** [reply]
Not for Thursday's update, but I will investigate changing this for the one after!

**Reply by Jdance-Media** [reply]
Hi, it might be worth adding this backup feature to map saving as well. I just tried to help someone who had their game crash/freeze up when they were saving and now their map data is corrupt with no backups available elsewhere. This is one of the exact use cases for having a file like that.

**Reply by DanielWillett** [reply]
Another sort of related thing you can do is not attempt to load save files if the version byte is higher than the current version, indicating that someone went from the preview branch back to the live branch.

Maybe supporting cancelling loading levels for important dat files and resetting less important dat files like editor tool configs, etc. would be a good addition.
